### PR TITLE
[java] Fix issue #17472 when using schemaMapping for models in collections, not compilable code with @Valid is generated

### LINKED ISSUE #17472: [BUG] generated code cannot be compiled (Java 21)
plugin version 7.2.0 generated code like below.

There are 2 problems:
1. code is not valid Java code (causes compiler error)
2. min and max length is not restriction for list itself, but for individual items in the list

Is there a workaround, how to switch this behavior off? ==> Possibly yes, not sure it will not disable.other older features -> `<useBeanValidation>false</useBeanValidation>`
(it was introduced here: https://github.com/OpenAPITools/openapi-generator/commit/809b3331a95b3c3b7bcf025d16ae09dc0682cd69)

```
class ExamplesApi.java
    default ResponseEntity<List<(min = 2, max = 2)String>> getLocalizations(

    //original code generated by 7.1.0:
    // default ResponseEntity<List<String>> getLocalizations(

class ExamplesApiDelegate.java
    default ResponseEntity<List<(min = 2, max = 2)String>> getLocalizations(
    //original code generated by 7.1.0:
    // default ResponseEntity<List<String>> getLocalizations(
```

example-service.json:
```
{
    "openapi" : "3.0.0",
    "info" : {
        "title" : "example-service",
        "description" : "undefined",
        "version" : "1.0.0"
    },
    "paths" : {
        "/localizations" : {
            "get" : {
                "description" : "example",
                "operationId" : "getLocalizations",
                "responses" : {
                    "200" : {
                        "description" : "OK",
                        "content" : {
                            "application/json" : {
                                "schema" : {
                                    "type" : "array",
                                    "items" : {
                                        "maxLength" : 2,
                                        "minLength" : 2,
                                        "type" : "string"
                                    }
                                }
                            }
                        }
                    }
                },
                "parameters" : [
                    {
                        "in" : "query",
                        "name" : "X-Correlation-ID",
                        "required" : false,
                        "schema" : {
                            "type" : "string"
                        }
                    },
                    {
                        "in" : "query",
                        "name" : "X-Message-ID",
                        "required" : false,
                        "schema" : {
                            "type" : "string"
                        }
                    }
                ],
                "tags" : [
                    "Examples"
                ]
            }
        }
    },
    "servers" : [
        {
            "url" : "http://localhost:8080/"
        }
    ],
    "tags" : [
        {
            "name" : "Examples",
            "description" : "Examples operations"
        }
    ]
}
```
plugin config

```
                <plugin>
                    <groupId>org.openapitools</groupId>
                    <artifactId>openapi-generator-maven-plugin</artifactId>
                    <version>7.2.0</version>
                    <executions>
                        <execution>
                            <goals>
                                <goal>generate</goal>
                            </goals>
                        </execution>
                    </executions>
                    <configuration>
                        <output>${project.basedir}/src/main/openapi</output>
                        <inputSpec>example-service.json</inputSpec>
                        <generatorName>spring</generatorName>
                        <library>spring-boot</library>
                        <ignoreFileOverride>.openapi-generator-ignore</ignoreFileOverride>
                        <supportingFilesToGenerate>OpenAPI2SpringBoot.java,OpenApiGeneratorApplication.java,RFC3339DateFormat.java,ApiUtil.java,</supportingFilesToGenerate>
                        <importMappings>SpecialId=org.example.common.SpecialId,</importMappings>
                        <typeMappings>special-id=SpecialId</typeMappings>
                        <configOptions>
                            <useSpringBoot3>true</useSpringBoot3>
                            <configPackage>org.example.service.configuration</configPackage>
                            <apiPackage>org.example.service.api</apiPackage>
                            <modelPackage>org.example.service.model</modelPackage>
                            <invokerPackage>org.example</invokerPackage>
                            <delegatePattern>true</delegatePattern>
                            <swaggerDocketConfig>false</swaggerDocketConfig>
                            <returnSuccessCode>true</returnSuccessCode>
                            <useTags>true</useTags>
                            <dateLibrary>java8</dateLibrary>
                            <hideGenerationTimestamp>true</hideGenerationTimestamp>
                        </configOptions>
                    </configuration>
                </plugin>

```

### PR DESCRIPTION
**Pull request by jpfinne**
Second PR to fix #17472

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 1]
```diff
@@ -1023,6 +1025,29 @@ public String getTypeDeclaration(Schema p) {
         return super.getTypeDeclaration(target);
     }
 
+    private String getTypeDeclarationForArray(Schema<?> items) {
```
Comment (arandth) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:1028]: Why do you call it "ForArray"? Isn't it valid for all Container types?

Comment (jpfinne) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:1028]: The method is used only once.  For OpenAPI this is an array element.

```
if (ModelUtils.isArraySchema(target)) {
  ...
            String typeDeclaration = getTypeDeclarationForArray(items);
```

What name would you use?

**Comment by jpfinne** [comment]
@wing328  Could you review (and hopefully merge) ?

**Comment by wing328** [comment]
@jpfinne thanks again for the PR. I tested with the spec and plugin config in the mentioned issue but still got errors

Are you able to repeat the issue in your local environment?

**Comment by arandth** [comment]
Hi @wing328 
I just again verified the issue, this time with the feature branch on this PR (yielding a 7.8.0-SNAPSHOT version of the generator).
I can confirm that for my problem the issue is solved with that version. My classes causing the problem are correctly generated.

For your information: here is the configuration section for the generator plugin (maybe you have differences that explain the different behavior):
                    <execution>
                        <id>exposed-provisioning-api-model-v1</id>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                        <configuration>
                            <inputSpec>${project.basedir}/src/main/resources/api/apigee-exposed/provisioning_api_model_v1.yaml</inputSpec>
                            <configOptions>
                                <useSpringBoot3>true</useSpringBoot3>
                                <dateLibrary>java8</dateLibrary>
                                <useBeanValidation>true</useBeanValidation>
                                <useOptional>false</useOptional>
                                <sourceFolder>.</sourceFolder>
                                <interfaceOnly>true</interfaceOnly>
                                <skipDefaultInterface>true</skipDefaultInterface>
                                <legacyDiscriminatorBehavior>true</legacyDiscriminatorBehavior>
                                <useTags>true</useTags>
                            </configOptions>
                            <generatorName>spring</generatorName>
                            <library>spring-boot</library>
                            <generateApis>false</generateApis>
                            <generateModels>true</generateModels>
                            <modelPackage>com.mycompany.subscriptionmodels.provisioning.v1</modelPackage>
                            <apiPackage>com.mycompany.subscriptionapis.provisioning.v1</apiPackage>
                            <generateModelDocumentation>false</generateModelDocumentation>
                            <generateModelTests>false</generateModelTests>
                            <generateSupportingFiles>false</generateSupportingFiles>
                            <skipIfSpecIsUnchanged>true</skipIfSpecIsUnchanged>
                            <inlineSchemaNameMappings>
                                _request_endpoint_post_request=RequestBase
                            </inlineSchemaNameMappings>
                            
                            
                            <schemaMappings>
                                FulfillmentRequest=com.mycompany.subscriptionmodels.provisioning.v1.FulfillmentRequest,
                                ProvisioningRequest=com.mycompany.subscriptionmodels.provisioning.v1.ProvisioningRequest
                                
                            </schemaMappings>
                        </configuration>
                    </execution>

Please note the commented "PendingRequest" class in schemaMapping is exactly the class causing the problem, as it has a List<FulfillmentRequest> as field (FulfillmentRequest is also in schemaMappings, as you can see)

**Comment by wing328** [comment]
@canadaduane thanks for testing the fix.

let's give this PR a try

**Pull request merged**